components/PackageList.tsx

```diff
@@ -8,13 +8,11 @@ type PackageListProps = {
     items: { name: string, version: string }[]
 }
 
-const PackageList: React.FC<PackageListProps> = (props) => {
-    const [packageList, setPackageList] = useState(props.items as { name: string, version: string }[]);
+const PackageList: React.FC<PackageListProps> = ({ items }) => {
+    const [packageList, setPackageList] = useState(items);
 
-    const onChange = (e: React.FormEvent<HTMLInputElement>) => {
-        const newValue = e.currentTarget.value;
-        setPackageList(props.items.filter(item => item.name.includes(newValue)));
-    }
+    const onChange = (e: React.FormEvent<HTMLInputElement>) =>
+        setPackageList(items.filter(item => item.name.includes(e.currentTarget.value)))
```

Comment:
This is a solid solution for filtering 👍 – alternatively in some cases we could to store the input value in a separate state or i.e. as a query parameter in the URL and then do filtering during the render:
```
const [filter, setFilter] = useState('');
const onChange = (e: React.FormEvent<HTMLInputElement>) => setFilter(e.currentTarget.value)

const packageList = filter ? items.filter(({ name }) => name.includes(filter)) : items;
```

Reply:
There's another sneaky benefit to Sauli's solution: As `PackageList.tsx` is currently written, it does not update correctly if the parent would update `items`. It would behave correctly only on the next time the filter is changed. In a data flow sense, you might think of `packageList` having a dependency to `items`, but there's nothing in this code that _reacts_ when that dependency changes.

In this particular app it _happens_ to work, since the property never changes, but in a client project this is a wtf-bug waiting to happen :)

Reply:
I get the idea of storing the filter state and yes I like it. What I am not that clear is the last row: 
```
const packageList = filter ? items.filter(({ name }) => name.includes(filter)) : items;
```
This seems to be called once. But should the packageList also change when the filter is changed?
So this makes me wonder if I get the meaning of rendering time correctly in this case. Is it so that the component is rendered on filter being changed? I would have thought not but...

Reply:
> This seems to be called once. But should the packageList also change when the filter is changed?
So this makes me wonder if I get the meaning of rendering time correctly in this case. Is it so that the component is rendered on filter being changed? I would have thought not but...

Good question! In short: the function that returns the component internals is referred generally as the `render` function. This is because if you make a React Class component, the function is actually named `render` 

The `render` function is automatically called everytime React notices that something has changed and the component needs to be refreshed. This can occur when e.g. `props` change or any of the internal `state` get changed.

Reply:
So based on this and on @wolfie comment above. If I understand correctly if the `items` in the props will change is it so that the whole `PackageList.tsx` component will be rendered meaning that the state of both packageList and filter will be reset depending which one is used with the `useState`.
Correct me if I'm wrong here:
- If I use `const [filter, setFilter] = useState('');` and the props changes, the value of filter is reset to '' so I still have to trigger the onChange event
- And same happens if I use packageList with useState. If the component is re-rendered because props have changed that value is again set to the whole items list and filtering is lost.

Reply:
> If I understand correctly if the items in the props will change is it so that the whole PackageList.tsx component will be rendered...

Correct! A simple way to see how often (and when) components are rendered is just to put a `console.log` inside the render function and check your console. It might be a surprising amount of times. I think of it as "react calls the function for as many times until it's sure that the component doesn't change anymore", but it's a bit more clever/magical than that.

> If I use `const [filter, setFilter] = useState('');` and the props changes, the value of filter is reset to '' so I still have to trigger the onChange event.

Actually, false, and is the very reason why `useState` exists - the empty string `''` value will be set only when the component is first created ("attached"). Any further updates to it will not empty the value - the value change only when any code calls the `setFilter` from now on.

> And same happens if I use packageList with useState. If the component is re-rendered because props have changed that value is again set to the whole items list and filtering is lost.

As above, nope. Instead of giving you the answer outright, you could try to figure this out yourself and get back to us when you think you got the issue. Or we can have a short call about it, if it helps.

I'd probably add a button to `pages/index.tsx` that toggles between two sets of `items` passed into `PackageList`, and then see how the filtering reacts to it. Protip, if you write e.g. `console.log({ items })` and `console.log({ packageList })` (note the curly braces) in `PackageList`'s render function, you'll clearly see what is what

Reply:
Back to this one, I get now the logic behind this. I also did the test by using different package lists  being set on the `PackageList` component from the `index.tsx` and yes I see the behaviour is what you describe and it is pretty much clear now.

My next question is what happens with the rendering lifecycle of a component when we move from one `NextPage` page to another. If I have a filter defined in the PackageList component (index.tsx page) and then navigate to the details page ([packageName].tsx)  and then get back to the `index.tsx` by clicking the link defined with the `Link` component, it seems so that the index.tsx page is rendered from scratch, all states preserved with useState are reset as I don't have the defined filter anymore.

So if I need to preserve states between pages in this example how do I do it?

Reply:
That is correct, in this case, the component is mounted, maybe updated and finally unmounted. If the component is then again mounted some other time, the `useState` will be reset.

State management across the application is another topic that we can get back to later. But at this point, the stupid simple answer is "it needs to be stored somewhere". Some of the popular methods that I can come up with off the top of my head is

- in the application root, usually with the help of other libraries, or even simply a huge `useState`
- some state can be preserved in the URL as query parameters (`/myPage/?filter=foo`)
- session/local state is one popular one as well
- in some cases you might want to store some of the session in the server, against some session token (e.g. a shopping cart)